Balance shards for an index more evenly across multiple data paths

dakrone opened the pull request:
When a node has multiple data paths configured, and is assigned all of the
shards for a particular index, it's possible now that all shards will be
assigned to the same path (see #16763).

This change keeps the same behavior around determining the "best" path for a
shard based on space, however, it enforces limits for the number of shards on a
path for an index from the single-node perspective. For example:

Assume you had a node with 4 data paths, where `/path1` has a tremendously high
amount of disk space available compared to the other paths. If you create an
index with 5 primary shards, the previous behavior would be to assign all 5
shards to `/path1`.

This change would enforce a limit of 2 shards to each data path for that
particular node, so you would end up with the following distribution:

- `/path1` - 2 shards (because it has the most usable space)
- `/path2` - 1 shard
- `/path3` - 1 shard
- `/path4` - 1 shard

Note, however, that this limit is only enforced at the local node level for
simplicity in implementation, so if you had multiple nodes, the "limit" for the
node is still 2, so assuming you had enough nodes that there was only 2 shards
for this index assigned to this node, they would still both be assigned to
`/path1`.

dakrone commented:
I forgot to mention that the limit of 2 shards above is not some arbitrary limit, it's `Math.floorDiv(shardCount, paths.length) + ((shardCount % paths.length) == 0 ? 0 : 1);`, it just happened to be 2 in example above.

ywelsch commented:
Is this an alternative to #26621?

dakrone commented:
@ywelsch I suppose it is now that I look at it, I hadn't looked at #26621 yet. Looking at it, it looks like it's a choice between a semi-random allocation and a deterministic one. Thoughts?

hub-cap commented:
Imho the right choice is the one that guarantees more shards allocated to the larger disk.

s1monw commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -845,6 +846,35 @@ public int getNodeLockId() {
+        final String indexUniquePathId = index.getUUID();
+        for (final NodePath nodePath : nodePaths) {
+            Path location = nodePath.indicesPath;
+            if (Files.isDirectory(location)) {
```
I am not sure why you do this check?

s1monw commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -845,6 +846,35 @@ public int getNodeLockId() {
+        for (final NodePath nodePath : nodePaths) {
+            Path location = nodePath.indicesPath;
+            if (Files.isDirectory(location)) {
+                try (DirectoryStream<Path> indexStream = Files.newDirectoryStream(location)) {
```
I wonder if it would be easier to do `if (Files.exists(location.resolve(indexUniquePathId))` then you don't need to iterate?

s1monw commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -845,6 +846,35 @@ public int getNodeLockId() {
+     * @return a map of NodePath to count of the shards for the index on that path
+     * @throws IOException if an IOException occurs
+     */
+    public ObjectLongHashMap<NodePath> shardCountPerPath(final Index index) throws IOException {
```
I am not a huge fan of returning a 3rd party lib class from a public interface. Can we rather return a `long[]` that index corresponds to the `NodePath[]` index in this class?

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/ShardPath.java:
```diff
@@ -189,22 +190,41 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
+                    FileStore fileStore = nodePath.fileStore;
+
+                    BigInteger usableBytes = BigInteger.valueOf(fileStore.getUsableSpace());
+                    assert usableBytes.compareTo(BigInteger.ZERO) >= 0;
```
can you please put a message in here?

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/ShardPath.java:
```diff
@@ -189,22 +190,41 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
+                    // Deduct estimated reserved bytes from usable space:
+                    Integer count = dataPathToShardCount.get(nodePath.path);
+                    if (count != null) {
+                        usableBytes = usableBytes.subtract(estShardSizeInBytes.multiply(BigInteger.valueOf(count)));
```
I don't understand this. if we already have `count` shards allocated it should be already taken into account in usable space? why would we multiply it here again?

dakrone replied:
This was a weighting that Mike originally added to make repeatedly assigned a shard to a given path weigh the pass less. I didn't change it (just used it verbatim).

I think it should be removed, however, in the subsequent work for this PR, since the space method will be rewritten from that perspective for that.

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/ShardPath.java:
```diff
@@ -189,22 +190,41 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
+                    if (pathToShardCount.get(nodePath) >= maxShardsPerPath) {
+                        // Too many shards for this index on this path, skip this path
+                        continue;
+                    } else if (bestPath == null || usableBytes.compareTo(maxUsableBytes) > 0) {
```
I think we should take the path with the most usable space and assign it to `bestPath` before we go into this optmization to be sure we take the biggest available and don't rely on iteration order.

s1monw replied:
In general, since we only find a spot for a single shard can't we just sort the list of paths by the number of shard for the index, prune by the required size and take the first one? wouldn't that be simpler algorithm wise?

dakrone replied:
Sure, I think we're just moving to the next step of the algorithm we discussed. It sounds like instead of getting this one in and then addressing that, we should just go for the final version. I'll update this PR to do that.

dakrone commented:
@elasticmachine test this please

dakrone commented:
@s1monw thanks for the feedback! I've updated this to have the same method we discussed (filtering and then sorting). I've also addressed your other comments, can you take another look?

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/ShardPath.java:
```diff
@@ -214,6 +240,30 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
         return new ShardPath(indexSettings.hasCustomDataPath(), dataPath, statePath, shardId);
     }
 
+    static boolean pathHasEnoughSpace(NodeEnvironment.NodePath path, BigInteger estimatedShardBytes) throws IOException {
```
I don't see where this is used? did I miss something or did you not use it?

dakrone replied:
Whoops yeah, this was a leftover, I'll remove it.

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/ShardPath.java:
```diff
@@ -189,22 +193,44 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
+                                }
+                                return cmp;
+                            })
+                        .collect(Collectors.toList());
```
getFirst() would be enough?

dakrone replied:
Yep, I'll use this.

dakrone commented:
Thanks @s1monw, I push two more commits based on your feedback

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/ShardPath.java:
```diff
@@ -189,23 +193,44 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
+                        .sorted((p1, p2) -> {
+                                int cmp = Long.compare(pathToShardCount.getOrDefault(p1, 0L), pathToShardCount.getOrDefault(p2, 0L));
+                                if (cmp == 0) {
+                                    // if the number of shards is equal, tie-break with the usable bytes
```
I wonder if we should tie-break based on the total number of shards on this path or do we want to fill up big disks first? I am not sure if that hot-spotting makes sense? maybe we first tie-break on # total shards and then on free space?

dakrone replied:
I agree with that, I'll work on that as a follow-up for this, thanks for the suggestion!

s1monw replied:
++

760515805 commented:
Hello,is there any progress on this issue,Because I have this problem now,my es version is 6.2.3.
